runtime/lua/vim/snippet.lua

```diff
@@ -207,9 +213,57 @@ function Session.new(bufnr, snippet_extmark, tabstop_data)
     end
   end
 
+  self:set_keymaps()
+
   return self
 end
 
+--- Sets the snippet navigation keymaps.
+---
+--- @package
+function Session:set_keymaps()
+  local function save_keymaps(keymaps, mode)
+    for _, keymap in ipairs(keymaps) do
+      if (keymap.lhs or ''):lower() == jump_forward_key then
+        self.tab_keymaps[mode] = keymap
+      elseif (keymap.lhs or ''):lower() == jump_backward_key then
+        self.shift_tab_keymaps[mode] = keymap
+      end
+    end
+  end
+
+  local function set(jump_key, direction)
+    vim.keymap.set({ 'i', 's' }, jump_key, function()
+      return vim.snippet.active({ direction = direction })
+          and '<cmd>lua vim.snippet.jump(' .. direction .. ')<cr>'
+        or jump_key
+    end, { expr = true, silent = true, buffer = self.bufnr })
+  end
+
+  save_keymaps(vim.api.nvim_buf_get_keymap(self.bufnr, 'i'), 'i')
+  save_keymaps(vim.api.nvim_buf_get_keymap(self.bufnr, 's'), 's')
+  set(jump_forward_key, 1)
+  set(jump_backward_key, -1)
+end
+
+--- Restores/deletes the keymaps used for snippet navigation.
+---
+--- @package
+function Session:restore_keymaps()
+  local function restore(keymap, lhs, mode)
+    if keymap then
+      vim.keymap.set(mode, lhs, keymap.rhs or keymap.callback, { buffer = self.bufnr })
```

Comment:
I this this is wrong since I'm not passing all the original keymap options, but if I use `keymap` as the 4th argument I get errors about `rhs` or `lhs` being invalid keys. I tried reconstructing the opts with code like `opts = { expr = keymap.expr }` but then I had errors like "replace_keycodes needs expr to not be null".

I appreciate ideas/alternatives here.

Reply:
You should use `mapset()` (with `nvim_buf_call`) for restoring a mapping.

Reply:
What happens if there's non-buffer-local keymap and I set the navigation mappings here? Will it just temporarily overwrite the global ones?

Reply:
Buffer-local mappings take priority over global mappings, so yes, there is no need to save and restore global mappings.